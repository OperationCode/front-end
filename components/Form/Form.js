import { connect } from 'formik';

export const Form = connect(({ formik: { handleReset, handleSubmit }, ...props }) => (
  <form onReset={handleReset} onSubmit={handleSubmit} noValidate {...props} />
));

Form.displayName = 'Form';
