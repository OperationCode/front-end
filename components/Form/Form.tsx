import { connect } from 'formik';
import type { DetailedHTMLProps, FormHTMLAttributes, FunctionComponent } from 'react';

const FormikConnectedForm = connect(({ formik: { handleReset, handleSubmit }, ...props }) => (
  <form onReset={handleReset} onSubmit={handleSubmit} noValidate {...props} />
)) as FunctionComponent<DetailedHTMLProps<FormHTMLAttributes<HTMLFormElement>, HTMLFormElement>>;

FormikConnectedForm.displayName = 'FormikConnectedForm';

export default FormikConnectedForm;
