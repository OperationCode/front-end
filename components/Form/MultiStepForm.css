.MultiStepForm {
  width: auto;
}

.containerFormErrorMessage {
  display: flex;
  justify-content: center;
}

.buttonGrouping {
  display: flex;
  flex-wrap: nowrap;
  justify-content: space-around;
  align-items: center;
  margin-left: auto;
  margin-right: auto;
  margin-top: 2rem;

  /*
   * Specific value just so outer edges of each button visually aligns with
   * the outer edges of inputs in our often-existing two-column forms
   */
  width: 92.5%;
}

.buttonGrouping button + button {
  margin-left: 1rem;
}

@media screen and (--small-viewport) {
  .buttonGrouping button {
    min-width: 100px;
  }
}

.forceButtonRight {
  justify-content: flex-end;
}
