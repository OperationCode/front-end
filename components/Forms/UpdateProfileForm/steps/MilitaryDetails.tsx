import { Field } from 'formik';
import * as Yup from 'yup';
import { validationErrorMessages } from 'common/constants/messages';
import { SelectSingle } from 'components/Form/Select/SelectSingle';
import { SelectMulti } from 'components/Form/Select/SelectMulti';
import type { OptionType } from 'components/Form/Select/ThemedReactSelect';

MilitaryDetails.title = 'Military Details';

MilitaryDetails.validationSchema = Yup.object().shape({
  branchOfService: Yup.array()
    .of(Yup.object().shape({ label: Yup.string(), value: Yup.string() }))
    .min(1, validationErrorMessages.required),
  payGrade: Yup.string().required(validationErrorMessages.required),
});

MilitaryDetails.initialValues = {
  branchOfService: [] as OptionType[],
  payGrade: '',
};

interface MilitaryDetailsProps {
  isSubmitting: boolean;
}

/**
 * There is also 'Military family member' and 'Not affiliated with the military'; however, these
 * selections are automatically chosen based on previous step answers.
 */
const branchOptions = [
  'U.S. Air Force',
  'U.S. Army',
  'U.S. Coast Guard',
  'U.S. Marine Corps',
  'U.S. Navy',
  'U.S. Space Force',
];

const payGradeOptions = ['E1-E5', 'E6-E9+', 'O1-O3', 'O4-O7+', 'WO1-3', 'WO3+'];

export function MilitaryDetails({ isSubmitting }: MilitaryDetailsProps) {
  return (
    <div className="flex flex-col gap-4">
      <Field
        name="branchOfService"
        label="Branch Of Service*"
        component={SelectMulti}
        options={branchOptions.map(option => ({ value: option, label: option }))}
        isDisabled={isSubmitting}
      />

      <Field
        name="payGrade"
        label="Pay Grade*"
        component={SelectSingle}
        options={payGradeOptions.map(option => ({ value: option, label: option }))}
        isDisabled={isSubmitting}
      />
    </div>
  );
}
