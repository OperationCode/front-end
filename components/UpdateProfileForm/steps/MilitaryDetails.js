import React from 'react';
import { bool } from 'prop-types';
import { Field } from 'formik';
import * as Yup from 'yup';
import { updateUser } from 'common/constants/api';
import { validationErrorMessages } from 'common/constants/messages';
import Input from 'components/Form/Input/Input';
import Select from 'components/Form/Select/Select';
import styles from './_steps.css';

MilitaryDetails.propTypes = {
  isSubmitting: bool,
};

MilitaryDetails.defaultProps = {
  isSubmitting: false,
};

MilitaryDetails.title = 'Military Details';

MilitaryDetails.validationSchema = Yup.object().shape({
  // TODO: use when() on these
  branchOfService: Yup.string()
    .nullable()
    .required(validationErrorMessages.required),
  yearsOfService: Yup.number()
    .nullable()
    .positive('Enter a number between 1 and 40.')
    .lessThan(41, 'Enter a number between 1 and 40.')
    .required(validationErrorMessages.required),
  payGrade: Yup.string()
    .nullable()
    .required(validationErrorMessages.required),
});

MilitaryDetails.initialValues = {
  branchOfService: '',
  yearsOfService: '',
  payGrade: '',
};

MilitaryDetails.submitHandler = async values => {
  await updateUser(values);
};

function MilitaryDetails({ isSubmitting }) {
  return (
    <>
      <div className={styles.row}>
        <Field
          className={styles.fullWidth}
          name="branchOfService"
          label="Branch Of Service*"
          component={Select}
          options={[
            {
              value: 'army',
              label: 'Army / Army Reserves / Army National Guard',
            },
            {
              value: 'navy',
              label: 'Navy / Navy Reserves',
            },
            {
              value: 'marine',
              label: 'Marine Corps / Marine Corps Reserves',
            },
            {
              value: 'airforce',
              label: 'Air Force / Air Force Reserves / Air National Guard',
            },
            {
              value: 'coastguard',
              label: 'Coast Guard / Coast Guard Reserves',
            },
          ]}
          disabled={isSubmitting}
        />

        <Field
          type="number"
          name="yearsOfService"
          label="Years Of Service*"
          component={Input}
          disabled={isSubmitting}
          min="1"
          max="40"
        />

        <Field
          type="text"
          name="payGrade"
          label="Pay Grade*"
          component={Input}
          disabled={isSubmitting}
        />
      </div>
    </>
  );
}

export default MilitaryDetails;
