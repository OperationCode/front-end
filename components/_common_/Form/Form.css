.form {
  display: flex;
  flex-direction: column;
  width: 300px;
  max-width: 100%;
  margin: 0 auto 2rem;
}

.form * {
  box-sizing: border-box;
}

.formControl ~ .formControl {
  margin-top: 16px;
}

.formControl input {
  background-color: #e6e6e6;
  color: #000000;
  border: none;
  font-size: 1.3em;
  padding: 0 16px;
  height: 50px;
  width: 100%;
}

.formControl input::placeholder {
  color: #000000;
}
