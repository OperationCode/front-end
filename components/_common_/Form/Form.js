import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import styles from './Form.css';

Form.propTypes = {
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.element]).isRequired,
  className: PropTypes.string,
};

Form.defaultProps = { className: '' };

function Form({ children, className }) {
  return (
    <form
      className={classNames({
        [`${styles.form}`]: true,
        [`${className}`]: className,
      })}
    >
      {children}
    </form>
  );
}

export default Form;
