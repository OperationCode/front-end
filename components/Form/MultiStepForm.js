/* eslint-disable unicorn/prevent-abbreviations */
import React, { useState } from 'react';
import { arrayOf, func, object } from 'prop-types';
import noop from 'lodash/noop';
import { Formik } from 'formik';
import { validStep } from 'common/constants/custom-props';
import Button from 'components/Button/Button';
import Form from 'components/Form/Form';
import Alert from 'components/Alert/Alert';
import ProgressIndicator from 'components/ProgressIndicator/ProgressIndicator';
import styles from './MultiStepForm.css';

MultiStepForm.propTypes = {
  // initialValues must be object where entire form's shape is described
  initialValues: object.isRequired,

  getErrorMessage: func.isRequired,
  onEachStepSubmit: func,
  onFinalSubmit: func.isRequired, // to be considered onSuccess
  steps: arrayOf(validStep).isRequired,
};

MultiStepForm.defaultProps = {
  onEachStepSubmit: noop,
};

export function MultiStepForm({
  steps,
  initialValues,
  onEachStepSubmit,
  onFinalSubmit,
  getErrorMessage,
}) {
  const [state, setState] = useState({
    errorMessage: '',
    stepNumber: 0,
  });

  const { errorMessage, stepNumber } = state;

  const isLastStep = () => {
    return stepNumber === steps.length - 1;
  };

  // We assume this method cannot be called on the last step
  const showNextStep = ({ setFieldTouched }) => {
    // TODO: Only untouch if value is '' or []
    const nextStepFieldNames = Object.keys(steps[stepNumber + 1].initialValues);
    nextStepFieldNames.forEach(fieldName => setFieldTouched(fieldName, false));
    setState(prevState => ({ ...prevState, stepNumber: prevState.stepNumber + 1 }));
  };

  // We assume this method cannot be called on the first step
  const showPreviousStep = ({ setFieldTouched }) => {
    // TODO: Only untouch if value is '' or []
    const previousStepFieldNames = Object.keys(steps[stepNumber - 1].initialValues);
    previousStepFieldNames.forEach(fieldName => setFieldTouched(fieldName, false));

    setState(prevState => ({ ...prevState, stepNumber: prevState.stepNumber - 1 }));
  };

  const handleError = error => {
    setState(prevState => ({ ...prevState, errorMessage: getErrorMessage(error) }));
  };

  const handleSubmit = async (values, formikBag) => {
    if (errorMessage) {
      // reset error message each submit
      setState(prevState => ({ ...prevState, errorMessage: '' }));
    }

    try {
      await onEachStepSubmit(values);

      const currentStepSubmitHandler = steps[stepNumber].submitHandler;
      await currentStepSubmitHandler(values);

      if (isLastStep()) {
        await onFinalSubmit(values);
        formikBag.setSubmitting(false);
        formikBag.resetForm();
      } else {
        formikBag.setSubmitting(false);
        showNextStep(formikBag);
      }
    } catch (error) {
      formikBag.setSubmitting(false);
      handleError(error);
    }
  };

  const CurrentStep = steps[stepNumber];
  const isFirstStep = stepNumber === 0;

  return (
    <Formik
      initialValues={initialValues}
      validationSchema={CurrentStep.validationSchema}
      onSubmit={handleSubmit}
      render={formikBag => (
        <Form className={styles.MultiStepForm} onSubmit={formikBag.handleSubmit}>
          <h3 className={styles.centerAligned}>{CurrentStep.title}</h3>

          <ProgressIndicator stepNumber={stepNumber} totalSteps={steps.length} />

          <CurrentStep {...formikBag} />

          <div className={styles.errorMessage}>
            {errorMessage && <Alert type="error">{errorMessage}</Alert>}
          </div>

          <div className={styles.buttonGrouping}>
            {!isFirstStep && (
              <Button
                theme="secondary"
                disabled={formikBag.isSubmitting}
                onClick={() => showPreviousStep(formikBag)}
                data-testid="Previous Step Button"
              >
                ← Previous
              </Button>
            )}

            {isLastStep() ? (
              <Button
                type="submit"
                theme="secondary"
                disabled={formikBag.isSubmitting}
                data-testid="Submit Multi-Step Form"
              >
                Submit ✓
              </Button>
            ) : (
              <Button
                type="submit"
                theme="secondary"
                disabled={formikBag.isSubmitting}
                fullWidth={isFirstStep}
                data-testid="Submit Step Button"
              >
                Next →
              </Button>
            )}
          </div>
        </Form>
      )}
    />
  );
}

export default MultiStepForm;
