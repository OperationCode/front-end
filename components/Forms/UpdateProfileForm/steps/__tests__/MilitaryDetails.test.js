import React from 'react';
import { Formik } from 'formik';
import { fireEvent, render, waitFor } from '@testing-library/react';
import OperationCodeAPIMock from 'test-utils/mocks/apiMock';
import createSnapshotTest from 'test-utils/createSnapshotTest';
import { KEY_CODES } from 'test-utils/identifiers';
import Form from 'components/Form/Form';

import MilitaryDetails from '../MilitaryDetails';

describe('UpdateProfileForm/Steps/MilitaryDetails', () => {
  afterEach(() => {
    OperationCodeAPIMock.reset();
  });

  it('should render in context of Formik', () => {
    createSnapshotTest(
      <Formik
        initialValues={MilitaryDetails.initialValues}
        validationSchema={MilitaryDetails.validationSchema}
      >
        <Form>
          <MilitaryDetails />
        </Form>
      </Formik>,
    );
  });

  it('should update user on submit', async () => {
    OperationCodeAPIMock.onPatch('auth/profile/').reply(200);

    const { container } = render(
      <Formik
        initialValues={MilitaryDetails.initialValues}
        validationSchema={MilitaryDetails.validationSchema}
        onSubmit={MilitaryDetails.submitHandler}
      >
        <Form>
          <MilitaryDetails />
        </Form>
      </Formik>,
    );

    const ReactSelect = container.querySelector('#react-select-branchOfService-input');

    fireEvent.blur(ReactSelect);
    fireEvent.keyDown(ReactSelect, KEY_CODES.DOWN_ARROW);
    fireEvent.keyDown(ReactSelect, KEY_CODES.DOWN_ARROW);

    fireEvent.keyDown(ReactSelect, KEY_CODES.ENTER);
    fireEvent.change(container.querySelector('input#yearsOfService'), {
      target: { id: 'yearsOfService', value: '3' },
    });
    fireEvent.change(container.querySelector('input#payGrade'), {
      target: { id: 'payGrade', value: 'E-5' },
    });
    fireEvent.submit(container.querySelector('form'));

    await waitFor(() => {
      expect(OperationCodeAPIMock.history.patch.length).toStrictEqual(1);
    });
  });
});
