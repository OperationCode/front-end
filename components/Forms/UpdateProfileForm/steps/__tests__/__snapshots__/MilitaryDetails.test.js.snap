// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UpdateProfileForm/Steps/MilitaryDetails > should render in context of Formik 1`] = `
<form
  noValidate={true}
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="row"
  >
    <input
      isMulti={true}
      label="Branch Of Service*"
      name="branchOfService"
      onBlur={[Function]}
      onChange={[Function]}
      options={
        [
          {
            "label": "Army / Army Reserves / Army National Guard",
            "value": "army",
          },
          {
            "label": "Navy / Navy Reserves",
            "value": "navy",
          },
          {
            "label": "Marine Corps / Marine Corps Reserves",
            "value": "marine",
          },
          {
            "label": "Air Force / Air Force Reserves / Air National Guard",
            "value": "airforce",
          },
          {
            "label": "Coast Guard / Coast Guard Reserves",
            "value": "coastguard",
          },
        ]
      }
      value={[]}
    />
    <div
      className="field"
      data-testid="INPUT"
    >
      <label
        className="text-themeSecondary text-sm"
        data-testid="LABEL"
        htmlFor="yearsOfService"
      >
        Years Of Service*
      </label>
      <div
        className="inputFeedbackGrouping"
        data-testid="INPUT_FEEDBACK_GROUPING"
      >
        <input
          className="Input"
          id="yearsOfService"
          max="40"
          min="1"
          name="yearsOfService"
          onBlur={[Function]}
          onChange={[Function]}
          type="number"
          value=""
        />
      </div>
    </div>
    <input
      label="Pay Grade*"
      name="payGrade"
      onBlur={[Function]}
      onChange={[Function]}
      options={
        [
          {
            "label": "E1-E5",
            "value": "E1-E5",
          },
          {
            "label": "E6-E9+",
            "value": "E6-E9+",
          },
          {
            "label": "O1-O3",
            "value": "O1-O3",
          },
          {
            "label": "O4-O7+",
            "value": "O4-O7+",
          },
          {
            "label": "WO1-3",
            "value": "WO1-3",
          },
          {
            "label": "WO3+",
            "value": "WO3+",
          },
        ]
      }
      value=""
    />
  </div>
</form>
`;
